In [1]:
# Ingnore future waring fro pandas 
import warnings
warnings.filterwarnings('ignore')
#import useful library

from tabulate import tabulate 
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

from plotly.subplots import make_subplots

In [2]:
pip install plotldy

Note: you may need to restart the kernel to use updated packages.


ERROR: Could not find a version that satisfies the requirement plotldy (from versions: none)
ERROR: No matching distribution found for plotldy


In [3]:
#import the dataset
autos = pd.read_csv('autos.csv',encoding = 'Latin-1')

autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

#### Obsercations 
-*our Dataset contains **20 columns** and **50,00 rows**. Most of the columns contain object of string data, a few columns contain intrgers.*
-*The price and odometer columns contain numeric values store as text.*
- *We have 5 columns with null values (vehicleType,gearbox, model,fuelType,notRepairedDamage) although these null values are less tgab 10% in each columns*
-*Columns with data information such as dateCrawled and dateCreated and lastSeen are saved as object/string data instead of dates.*
-* 6 columns have string data recorded in German language. These columns ase - seller, offerType,vehicleType,gera-box,fueltype and notRepair Damage.*
-*Rather than pythons's perferred snakecase, the columns names follow camelcase convention*

As we explore further , we will try to correct these occurences, so they don't cause us problems later.

## Data cleaning 
## 1.cleaning column names 

we will start by converting the columns fro camelcase to snakecase.let's print these columns to identify the ones we need rename.


In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
               'vehicle_type', 'registration_year', 'gear_box', 'powerPS', 'model',
               'odometer', 'registration_month', 'fuel_type', 'brand',
               'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
               'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

### detele the columns 
We will detele the columns which only have few unique values.Due to the face that those columns are hard to analysis anything valueable.

In [8]:
# loop through each columns and print the frequency of rach unique value.

for items in ['seller','offer_type', 'ab_test', 'gear_box', 'unrepaired_damage']:
  print(autos[items].value_counts())
  print('                   ')

privat        49999
gewerblich        1
Name: seller, dtype: int64
                   
Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64
                   
test       25756
control    24244
Name: ab_test, dtype: int64
                   
manuell      36993
automatik    10327
Name: gear_box, dtype: int64
                   
nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64
                   


In [9]:
autos['num_pictures'].value_counts()

0    50000
Name: num_pictures, dtype: int64

In [10]:
print(autos.shape)
autos = autos.drop(['seller','offer_type','num_pictures'],axis=1)
print(autos.shape)

(50000, 20)
(50000, 17)


## 2.translating non-english words 

In the previous observation, we manually interpreted the meaning of the terms in the seller and offer type columns .Reapting this processes for all affected colmns will be cumbersome in the long run.

In total, there were six columns recored in ferma. Howver, we had already dropped two of the affected columns. we well explore tge remaining four in details:

In [11]:
german_columns = ['vehicle_type','gear_box','fuel_type','unrepaired_damage']

# print the unique german words in each columns of interest

for column in german_columns:
  print('')
  print(autos[column].unique())



['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']

['manuell' 'automatik' nan]

['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']

['nein' nan 'ja']


We cam use this imforamtion to build a dictionary of translation for each german word.Then,use a function  `convert_language()` to iterate through each of the affected columns, while mapping the right English words from our dictionary. The `series.map()` method will handle the translation/mapping process for us:

In [12]:
def translate_column(df,column,value_dict):
  df[column] = df[column].map(value_dict)

# A dictionary of translations
translations ={# vehicle type terms
               'bus':'bus',
               'limousine':'limousine',
               'kleinwagen': 'small car',
               'kombi': 'combo',
               'coupe':'coupe',
               'suv':'suv',
               'cabrio': 'cabrio',
               'andere': 'other',
               'cabrio': 'convertible',
               # gear box terms
               'manuell': 'manual',
               'automatik': 'automatic',
               # fuel type terms
               'lpg':'lpg',
               'benzin': 'diesel',
               'diesel':'diesel',
               'cng':'natural gas',
               'hybrid':'hybrid',
               'elektro':'electric',
               'andere':'other',
               # unrepaired damage terms
               'nein':'no',
               'ja':'yes'       
}



Let's translate each of these columns. We will also print out the value counts before and after translation to ensure that the process ran adequately, without compromising our data:

In [13]:

for item in german_columns:
  print('Before translating','\n')

  print(autos[item].value_counts(dropna=False))
  print('\n','TRANSKATTING WORDS>>>>')
  print('-'*30)
  translate_column(autos,item,translations)
  print(autos[item].value_counts(dropna=False))
  print('_________________________')



Before translating 

limousine     12859
kleinwagen    10822
kombi          9127
NaN            5095
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

 TRANSKATTING WORDS>>>>
------------------------------
limousine      12859
small car      10822
combo           9127
NaN             5095
bus             4093
convertible     3061
coupe           2537
suv             1986
other            420
Name: vehicle_type, dtype: int64
_________________________
Before translating 

manuell      36993
automatik    10327
NaN           2680
Name: gear_box, dtype: int64

 TRANSKATTING WORDS>>>>
------------------------------
manual       36993
automatic    10327
NaN           2680
Name: gear_box, dtype: int64
_________________________
Before translating 

benzin     30107
diesel     14567
NaN         4482
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64


#### Observations
- *The German words have been converted to English, and the value counts remained the same after translation. We can conclude that the translation was executed correctly without compromising our data.*

## 3. Reformating Numeric Data

We previously identified some object types that should have been saved as numeric; like the `price` and `odometer` columns. For each of these columns, we will remove the non-numeric characters, convert the values to numeric, then use the `DataFrame.rename()` method to assign descriptive names.

The `price` columns contains non-numeric characters like `','` and `'$'` non-numeric characters, while the The `odometer` column has `','` and `'km'` as its non-numeric characters.

In [14]:
autos['price'] = autos['price'].str.replace(',','').str.replace('$','').astype(int)
autos.rename({'price':'usd_price'},inplace=True,axis=1)


autos['usd_price'].head(5)

0    5000
1    8500
2    8990
3    4350
4    1350
Name: usd_price, dtype: int32

In [15]:
autos['odometer'] = autos['odometer'].str.replace(',','').str.replace('km','').astype(int)
autos.rename({'odometer':'odometer_km'},inplace=True,axis=1)


autos['odometer_km'].head(5)

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int32

##4 .Detect the numeric outlieries

Let's continue by explore the data for outliers.Specifically, we are looking for data that doesn't not look right,espically in the numeric columns .We will analysis these columns using the five number summary (min,Q!,median Q2,max) then look fot any valyue thay look unrealistically high or low.

In [16]:
autos[['usd_price','odometer_km']].describe()

usd_price    odometer_km
count  5.000000e+04   50000.000000
mean   9.840044e+03  125732.700000
std    4.811044e+05   40042.211706
min    0.000000e+00    5000.000000
25%    1.100000e+03  125000.000000
50%    2.950000e+03  150000.000000
75%    7.200000e+03  150000.000000
max    1.000000e+08  150000.000000

#### Observations

- ***Odometer readings***
    - *The median and 75th percentile value are the same as the maximum value (150,000km). The 25th percentile is set at around 125,000km.*
    - *This indicates that a reading of 150,000km is not an outlier, since 75% of the listed cars are between 125,000 to 150,000km in mileage.*

- ***Price***
    - *The maximum price is 100,000,000USD. This varies significantly from the median and 75th percentile values of 2,950 and 7,200USD respectively. The standard deviation is also considerably high (about 480,000 USD).*
    - *The minimum price of 0 is unusual and could affect our analysis.* 

We can further explore `odometer_km` by displaying its sorted value counts:

In [17]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

#### Additional Observations

- *The least mileage observed is 5000km rather than 0km. This is expected because the cars listed on this platform are used cars.*
- *Higher Mileage cars (above 50,000km) are more common on the listing than lower mileage ones.*
- *Over 30,000 (~65%) cars have odometer readings of 150,000. This could imply that there is an upper mileage limit of 150,000km for cars accepted for listing.*

Next, we will explore `usd_price` and attempt to remove its outliers:

In [18]:
autos['usd_price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: usd_price, dtype: float64

## 5. Removing Price Outliers

As observed earlier, the `25th`, `50th` and `75th` percentile of the prices are `1,100`, `2,950` and `7,200`USD. We will use a selection method to create a **“data fence”** from our 25th to 75th percentiles. Any data beyond this fence will be considered an outlier.

We already have enough information to calculate our interquartile range `IQR`. We will use the equation below to determine our upper and lower price fences:

```
    Interquartile Range (IQR) = 75th Percentile - 25th Percentile
    Lower fence = 25th Percentile - 1.5*(IQR)
    Upper fence = 75th Percentile + 1.5*(IQR)
    
```
Lets define all these boundaries in the code cell below:

In [19]:
price_25,price_75 = (1100,7200)

IQR = price_75-price_25
lower_fence = price_25-(1.5*IQR)
upper_fence = price_75+(1.5*IQR)

print(IQR)
print(lower_fence)
print(upper_fence)


6100
-8050.0
16350.0


Fences may extend beyond the data sometimes, so it is totally fine if our lower fence is negative. We should be mindful, however, that a minimum price value of `0` can also affect our analysis. In this case, it is advisable to eliminate the entries with zero prices too.

Before we proceed, let's take a look at how many listings are recorded at zero price:

In [20]:
autos['usd_price'].value_counts().head()

0       1421
500      781
1500     734
2500     643
1000     639
Name: usd_price, dtype: int64

Of the 50,000 rows in our dataset, 1,421 records `(3%)` are listed at zero price. Since this percentage is very low, we can safely remove these entries. We will set our lower fence at `1` to exclude zero prices:

In [21]:
autos = autos[autos['usd_price'].between(1,upper_fence)]

In [22]:
autos['usd_price'].describe()

count    44795.000000
mean      4089.433620
std       3841.429247
min          1.000000
25%       1150.000000
50%       2700.000000
75%       5999.000000
max      16350.000000
Name: usd_price, dtype: float64

#### Observations

- *Our dataset has been reduced to 44,795 rows in total. The 25th, 50th and 75th percentiles are close to their initial values before cleaning.*
- *The standard deviation (3,841USD) is considerably lower than the value of 480,000USD recorded before.*
- *16,350USD is a better approximation for the maximum price than the previous value of 100,000,000USD, considering that our 75th percentile is around 6,000USD.*

Let's now explore the date columns to understand the date range our data covers

## Exploring the Date Columns

There are 5 columns in our dataset that should represent date values. Some of these columns were created by the crawler, while some were obtained from the website. We can differentiate by referring to the information below:

```
- `date_crawled` - added by the crawler
- `last_seen` -  added by the crawler
- `ad_created` - obtained from the website
- `registration_month` - obtained from the website
- `registration_year`- obtained from the website

```

`date_crawled`, `last_seen`, and `ad_created` are currently identified as object types. To understand them quantitatively, we need to convert them to `datetime` objects.

`registration_month` and `registration_year` are represented as numeric values. Using the `.describe()` method, we can understand their distribution, without any conversion.

Let's first understand how the values in the three string columns are formatted.

In [23]:
autos[['date_crawled', 'ad_created', 'last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

#### Observations

- *It appears that the first 10 characters in each column represent the date in the yy-mm-dd format (e.g. 2016-03-12). The other characters represent hours:minutes:seconds*

To understand the distribution of the dates, we can extract just the relevant values using a string method. We can use `Series.str[:10]` to select the first 10 characters in each column.

### 1. Date Crawled

In [24]:
print(autos['date_crawled'].str[:10].unique().size)

date_crawled = (
                autos['date_crawled']
                    .str[:10]
                    .value_counts(normalize=True,dropna=False)
                    .round(3)
                    .sort_index())



#use tabulate to print a preety table

print(tabulate(date_crawled.to_frame(),headers=['Date','% crawled'],tablefmt='psql'))



# viusalize data 

fig = px.line(x = date_crawled.index,y=date_crawled.values,template='plotly_white',
title = 'Ads Crawled Per Day (March-April 2016)')

fig.update_xaxes(nticks=34, title='Date',showgrid=False, ticks='outside', linecolor='black')
fig.update_yaxes(title='Ads crawled (% of Total)',showgrid=False, linecolor='black')
fig.update_layout(font_size=12,font_family = 'Arial')


34
+------------+-------------+
| Date       |   % crawled |
|------------+-------------|
| 2016-03-05 |       0.026 |
| 2016-03-06 |       0.014 |
| 2016-03-07 |       0.036 |
| 2016-03-08 |       0.034 |
| 2016-03-09 |       0.033 |
| 2016-03-10 |       0.033 |
| 2016-03-11 |       0.032 |
| 2016-03-12 |       0.037 |
| 2016-03-13 |       0.016 |
| 2016-03-14 |       0.037 |
| 2016-03-15 |       0.034 |
| 2016-03-16 |       0.03  |
| 2016-03-17 |       0.032 |
| 2016-03-18 |       0.013 |
| 2016-03-19 |       0.034 |
| 2016-03-20 |       0.038 |
| 2016-03-21 |       0.037 |
| 2016-03-22 |       0.033 |
| 2016-03-23 |       0.032 |
| 2016-03-24 |       0.029 |
| 2016-03-25 |       0.032 |
| 2016-03-26 |       0.033 |
| 2016-03-27 |       0.031 |
| 2016-03-28 |       0.035 |
| 2016-03-29 |       0.034 |
| 2016-03-30 |       0.034 |
| 2016-03-31 |       0.032 |
| 2016-04-01 |       0.033 |
| 2016-04-02 |       0.035 |
| 2016-04-03 |       0.039 |
| 2016-04-04 |       0.036 |
| 2016-04-0

#### Observations
- *It appears the crawler collected information from the site everyday between March 5, 2016 and April 7, 2016, a period of 34 days.*
- *Some occassional drops can be observed on some weekend days in March (6th, 13th, 18th); and the last 3 days in April. However, the number of ads crawled seem evenly distributed overall.*

### 2. Ad Created

In [25]:
ad_created = (autos['ad_created']
                .str[:10]
                .value_counts(normalize=True,dropna=False)
                .sort_index())

#examine the earliset and last ad created dates.

print('Earilest Date',ad_created.head(1))
print()
print('Last Date',ad_created.tail(1))

# visualize data

fig = px.line(x = ad_created.index,y = ad_created.values,template='plotly_white',
                title='Ads Created Per Day(August 10 - April 7 2016)')
fig.update_xaxes(title='Date',showgrid=False)
fig.update_yaxes(title='Ads created(% of Total)',showgrid=False, showline=True)

# label the date where the most noabel and increase occured 
fig.add_annotation(x='2016-03-05',y=0.02302195, text= 'March 5',showarrow=True, arrowhead=1,
                   bordercolor="#c7c7c7", borderwidth=2,borderpad=4, bgcolor="#ff7f0e")

fig.update_layout(font_size = 12, font_family='Arial')

Earilest Date 2015-08-10    0.000022
Name: ad_created, dtype: float64

Last Date 2016-04-07    0.001183
Name: ad_created, dtype: float64


#### Observations

- *It appears that we are dealing with listings created between August 10, 2015, and April 7, 2016.* 
- *Earlier, there were **very few** listings on the site. This pattern changed, especially on March 5, 2016, when we noticed a massive growth in listings.  The number of listings has remained higher since then.*
- *Interestingly, March 5, 2016, was also the exact date when the crawler crawled the first group of ads from the website.*

### 3. Last Seen

In [26]:
last_seen = (autos['last_seen']
            .str[:10]
            .value_counts(normalize=True,dropna=False)
            .round(3)
            .sort_index()
)

# pretty print date wtih tabualte 

print(tabulate(last_seen.to_frame(),headers=['Date','last_seen%'],tablefmt='psql'))

fig = px.line(x=last_seen.index,y=last_seen.values,template='plotly_white',
title='Ads Crawled Per Day(March - April 2016)')
fig.update_xaxes(nticks=34, title='Date',showgrid=False, ticks='outside', linecolor='black')
fig.update_yaxes(title='Ads crawled (% of Total)',showgrid=False, linecolor='black')
fig.update_layout(font_size = 12, font_family='Arial')


+------------+--------------+
| Date       |   last_seen% |
|------------+--------------|
| 2016-03-05 |        0.001 |
| 2016-03-06 |        0.005 |
| 2016-03-07 |        0.006 |
| 2016-03-08 |        0.008 |
| 2016-03-09 |        0.01  |
| 2016-03-10 |        0.011 |
| 2016-03-11 |        0.013 |
| 2016-03-12 |        0.025 |
| 2016-03-13 |        0.009 |
| 2016-03-14 |        0.013 |
| 2016-03-15 |        0.016 |
| 2016-03-16 |        0.017 |
| 2016-03-17 |        0.029 |
| 2016-03-18 |        0.008 |
| 2016-03-19 |        0.016 |
| 2016-03-20 |        0.021 |
| 2016-03-21 |        0.021 |
| 2016-03-22 |        0.022 |
| 2016-03-23 |        0.019 |
| 2016-03-24 |        0.02  |
| 2016-03-25 |        0.02  |
| 2016-03-26 |        0.017 |
| 2016-03-27 |        0.016 |
| 2016-03-28 |        0.022 |
| 2016-03-29 |        0.023 |
| 2016-03-30 |        0.025 |
| 2016-03-31 |        0.024 |
| 2016-04-01 |        0.023 |
| 2016-04-02 |        0.025 |
| 2016-04-03 |        0.026 |
| 2016-04-

#### Observations

- *Since we know that the crawler also added the last seen information, it makes sense that the last seen column shows the same 34-day time period as the date crawled (March 5, 2016 and April 7, 2016).*
- *The number of last seen ads were relatively uniform throughout the crawling period. However, it appears that the crawler noticed an increase in ad disappearances in the final few days (April 5 - April 7, 2016).*
- *This may give bits of insight into the crawlers activity. We can assume the crawlers daily routine: The crawler visits the site, collects information on different ads, then records the date crawled and last seen information for various ads.*

### 4. Registration Year

In [27]:
autos['registration_year'].describe()

count    44795.00000
mean      2003.97339
std         74.81955
min       1000.00000
25%       1999.00000
50%       2003.00000
75%       2007.00000
max       9999.00000
Name: registration_year, dtype: float64

#### Observations

- ***The registration year column contains some unexpected values:***
    - *The minimum value is 1000, this is long before 1885 when cars were invented.*
    - *The maximum value is 9999, many years into the future. We are also aware that the listings were last seen by the crawler in 2016. A car can't be first registered after its listing has been seen.* 

Determining the earliest valid year can be challenging. However since we know that cars were invented in 1885, we can use the first few decades of 1900s as our lower cut off point. At the higher end, any record with registration year above 2016 is definitely inaccurate. 

Before we eliminate these rows, let's count the number of listings with cars that fall outside the 1900 - 2016 registration year and see if it's safe to remove these rows entirely without impacting our analysis too much.

In [28]:
invalid_rows = (~(autos['registration_year'].between(1900,2016))).sum()
total_row = autos['registration_year'].shape[0]
percentage_invalid = invalid_rows/total_row

print('The invaild years constitute {:.1f}% of the total data'.format(percentage_invalid*100))

The invaild years constitute 4.1% of the total data


 these years constitute only 4.1% of our total data. We can safely eliminate these records from our analysis:

In [29]:
autos = autos[autos['registration_year'].between(1900,2016)]

years = (autos['registration_year'].value_counts(normalize=True).sort_values(ascending=False))


fig = px.histogram(x =years.index,y=years.values,nbins=30,title='Distributions of cars registered between 1900 and 2016')

fig.update_traces(marker_color='#ff6347')
fig.update_xaxes(nticks=20,title='Years', showline=True, linewidth=2, linecolor='black', ticks='outside')
fig.update_yaxes(title='% of Registrations')
fig.update_layout(font_size = 12, font_family='Arial')

fig.show()

#### Observations

- *It appears that most of the vehicles were registered between 1990 and 2010. Only a small proportion of listed cars were registered before 1990.*

### 5. Registration Month

In [30]:
autos['registration_month'].describe()

count    42959.000000
mean         5.791918
std          3.691957
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

#### Observations:

- *Since the lowest value is set at 0 (presumably denoting January), we would expect the highest value to be at 11 (denoting December). However the highest value we see here is 12. This is quite unexpected.*

We will probe into this further:

In [31]:
autos['registration_month'].value_counts(normalize=True).sort_index()

0     0.092344
1     0.065760
2     0.060500
3     0.103308
4     0.082846
5     0.083405
6     0.087479
7     0.077679
8     0.064527
9     0.068624
10    0.074932
11    0.068111
12    0.070486
Name: registration_month, dtype: float64

#### Observations

- **_This looks a bit more challenging:_**
    - *The months are evenly distributed. Hence it would not be a good idea to remove any month in favor of the others*
    - *No month is erroneously skipped, which further complicates what 0 and 12 are supposed to represent.*

Since it is unclear where the error is from, we will drop the `registration_month` column as it could compromise our analysis:

In [32]:
autos = autos.drop('registration_month',axis=1)
autos.shape

(42959, 16)

## Data Analysis

## Exploring Price by Brand

Lets explore the unique values in the brand column to identify the most common brands in the used car listings:

In [33]:
# compute the relative proportions of each brands

brands = autos['brand'].value_counts(normalize=True)

# Extract the top-10 for visualization purpoes
top_10 = brands.head(10).round(2)

print(tabulate(brands.to_frame(),headers=['Brands','% of Listings'],tablefmt='psql'))

print(top_10)

fig = px.bar(x=top_10.index, y=top_10.values, text=top_10.values,
             template='plotly_white', title='Most Common Brands on the Listings (Top 10)')

fig.update_xaxes(title='Brand Name', showline=True, linewidth=2, linecolor='black')
fig.update_yaxes(title='Fraction of Total Listings',showgrid=False, showline=True, showticklabels=False)
fig.update_layout(font_size = 12, font_family='Arial')

+----------------+-----------------+
| Brands         |   % of Listings |
|----------------+-----------------|
| volkswagen     |     0.214763    |
| opel           |     0.115203    |
| bmw            |     0.103541    |
| mercedes_benz  |     0.0881305   |
| audi           |     0.0775623   |
| ford           |     0.0730929   |
| renault        |     0.0507693   |
| peugeot        |     0.0321702   |
| fiat           |     0.0277474   |
| seat           |     0.019088    |
| skoda          |     0.0169929   |
| mazda          |     0.0160851   |
| nissan         |     0.0157825   |
| smart          |     0.0153635   |
| citroen        |     0.0150376   |
| toyota         |     0.0133616   |
| hyundai        |     0.0105449   |
| volvo          |     0.00933448  |
| mitsubishi     |     0.00870598  |
| sonstige_autos |     0.00849647  |
| honda          |     0.00833353  |
| mini           |     0.00775158  |
| alfa_romeo     |     0.00700668  |
| kia            |     0.00696012  |
|

#### Observations:

- *The most popular brands are Volkswagen (21%). Opel (about 12%), BMW (10%), Mercedes Benz (9%) and Audi (8%). Ford (7%) and Renaut (5%) follow behind respectively.* 
- *If we summed these percentages, we would find out that these top 7 brands constitute about 72% of the total population of listed brands. Other brands constitute a combined 28%.* 
- *The top 5 brands are German (Volkswagen, Opel, BMW, Mercedes and Audi). They constitute about 50% of all car listings*

We will limit our analysis to brands representing more than 5% of the total. Interestingly, only the top 7 brands meet this criterion.

In [34]:
# extract the names of the top7 brands 

top_brands = autos['brand'].value_counts().head(7).index

top_brands

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford',
       'renault'],
      dtype='object')

Next,let's compute and compare the average prices of cars in the top 7 brands

In [35]:
aggregate_price = {}

for brand in top_brands:
    price = autos['usd_price'][autos['brand']==brand]
    mean_price= price.mean()
    aggregate_price[brand] = int(mean_price)

aggregate_price

{'volkswagen': 4183,
 'opel': 2712,
 'bmw': 5649,
 'mercedes_benz': 5259,
 'audi': 5703,
 'ford': 2944,
 'renault': 2281}

In [36]:
# convert dictionary to a pandas series for easy printing and manipulatrion

price_series = pd.Series(aggregate_price)
price_series = price_series.sort_values(ascending=False)
print(tabulate(price_series.to_frame(), headers=['Brands', 'Average price'], tablefmt='psql'))

fig = px.bar(x=price_series.index, y=price_series.values, text=price_series.values,
             template='plotly_white', title='Average Price Across the Top 7 Brands')

fig.update_xaxes(title='Brand Name', showgrid=False, showline=True, linewidth=2, linecolor='black')
fig.update_yaxes(title='Average price in USD',showgrid=False, showline=True, showticklabels=True,
                ticks='outside', tickprefix='$')
fig.update_layout(font_size = 12, font_family='Arial')
fig.update_traces(textposition='outside')


+---------------+-----------------+
| Brands        |   Average price |
|---------------+-----------------|
| audi          |            5703 |
| bmw           |            5649 |
| mercedes_benz |            5259 |
| volkswagen    |            4183 |
| ford          |            2944 |
| opel          |            2712 |
| renault       |            2281 |
+---------------+-----------------+


#### Observations

- *Asides from Opel, the German brands (Volkswagen, BMW, Mercedes and Audi) seem more expensive on average than the American or French ones.*
- *The average price for Volkwagen seems to be at the middle point of all these prices, offering an interesting balance between its popularity and price.*

## The Top 7 Brands - Exploring Mileage

For the top 7 brands, we will use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. Rather than display both aggregated series objects and visually compare them, we will combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly.

We have already converted our `aggregate_price` dictionary into the Pandas series object - `price_series` using the `pd.Series()` constructor. Next we will convert the series object into a Dataframe using the `pd.DataFrame()` constructor:

In [37]:
price_df = pd.DataFrame(data = price_series,columns=['mean_price_usd'])
price_df

mean_price_usd
audi                     5703
bmw                      5649
mercedes_benz            5259
volkswagen               4183
ford                     2944
opel                     2712
renault                  2281

In [38]:
aggregate_mileage={}
for brand in top_brands:
    mileage = autos['odometer_km'][autos['brand']==brand]
    mean_mileage= mileage.mean()
    aggregate_mileage[brand] = int(mean_mileage)

mileage_series = pd.Series(aggregate_mileage)
mileage_df = pd.DataFrame(data = mileage_series,columns = ["mean_mileage_km"])
mileage_df

mean_mileage_km
volkswagen              132813
opel                    130500
bmw                     138833
mercedes_benz           138411
audi                    139701
ford                    127004
renault                 128892

Finally, we will join the average price and mileage dataframes together to enable us compare their results at a glance. One way we can do this is by assigning `mileage_series` to a new column in the price dataframe:

In [39]:
price_df['mean_mileage_km'] = mileage_series
price_df = price_df.sort_values(by='mean_mileage_km',ascending=False)
price_df

mean_price_usd  mean_mileage_km
audi                     5703           139701
bmw                      5649           138833
mercedes_benz            5259           138411
volkswagen               4183           132813
opel                     2712           130500
renault                  2281           128892
ford                     2944           127004

In [40]:
# Visualize results
fig = make_subplots(rows=1, cols=2, horizontal_spacing=0.2)

fig.add_trace((go.Bar(y=price_df.index, x=price_df['mean_price_usd'], orientation='h',
                      name='mean_price_usd'
                     )
              ), row=1, col=1)

fig.add_trace((go.Bar(y=price_df.index, x=price_df['mean_mileage_km'], orientation='h',
                      name='mean_mileage_km'
                     )
              ), row=1, col=2)

fig.update_layout(template = 'plotly_white', title_text= 'Average Price and Mileage Across the Top 7 Brands',
                 height=500, font_size = 12, font_family='Arial')

fig.update_yaxes(showline=True, linewidth=1, linecolor='black', mirror=True)
fig.update_xaxes(showline=True, linewidth=1, linecolor='black', mirror=True)
fig.update_xaxes(tickprefix='$',row=1, col=1, title='Average Price')
fig.update_xaxes(row=1, col=2, title='Average Mileage (km)')

#### Observations

- *We would expect car prices to fall as mileage increases. However this is not the case here as we notice the more expensive vehicles having higher mileage, and the cheaper ones recording lower mileages.*
- *We see Volkswagen at the median of the mileage values again. Could its popularity on the listings be perhaps attributable to its balance in price and mileage?*

